components/ws-manager/pkg/manager/status.go

```diff
@@ -591,11 +587,8 @@ func extractFailure(wso workspaceObjects) (string, *api.WorkspacePhase) {
 					return fmt.Sprintf("container %s ran with an error: exit code %d", cs.Name, terminationState.ExitCode), &phase
 				}
 			} else if terminationState.Reason == "Completed" {
-				if wso.IsWorkspaceHeadless() {
-					// default way for headless workspaces to be done
-					return "", nil
-				}
-				return fmt.Sprintf("container %s completed; containers of a workspace pod are not supposed to do that. Reason: %s", cs.Name, terminationState.Message), nil
+				// container terminated successfully - this is not a failure
```

Comment:
The original intent with this branch was to cover the case that a regular workspace stops without being deleted. An unintentional stop, so to speak. That said, the failure mode is actually nicer if your workspace just stops compared to get getting an inactionable error message ("containers of a workspace pod are not supposed to do that" 🤷).

The `failed` condition on workspace status indeed isn't stable (but should be). ws-manager-bridge should protect us against that by not resetting that condition if it ever was set. Because we have nothing like an at-least-once delivery semantic, we might just miss the "failed" message though.